Explain some more commit stuff

ericonr opened the pull request.

ericonr committed:
tutorial: mention atomic commits.

ericonr committed:
tutorial: add slide for commit messages.

ericonr committed:
tutorial: call out git-diff when staging.

It's easy to accidentally stage something you didn't want to include in
a commit, using git diff --staged allows you to check for that before
getting to the commit stage.

henriquesimoes commented on git-tutorial.tex:
```diff
@@ -76,7 +76,8 @@ \subsection{Staging}
     \item Avoid using \texttt{git add .} as it may inadvertently stage undesirable files. \texttt{git add --update} will only add files which are already tracked by git, which is usually closer to the desired effect;
     \item You can stage parts of a file with \texttt{git add -p} \textit{file};
-    \item To unstage a file, use the \texttt{git reset} \textit{file} command.
+    \item To unstage a file, use the \texttt{git reset} \textit{file} command;
```
Not actually related to this PR, but isn't the new `git-restore(1)` more appropriate for that?

That's what is recommended by the Git itself in its advice messages nowadays.

ericonr replied:
Isn't `git restore` supposed to replace `git checkout`? In this case, `reset` is being used only to unstage a file, not to discard its changes.

ericonr replied:
I didn't know there was `git restore --staged`

I don't love it, because forgetting the flag might make one lose data, when `git reset` can be run without any flags.

henriquesimoes commented on git-tutorial.tex:
```diff
@@ -87,9 +87,10 @@ \subsection{Staging}
+    \item After staging your modifications, it's time to commit the changes;
+    \item Commit messages ideally should contain a title, following a blank line, and a more \textbf{in-depth explanation} of the changes;
+    \item A commit should cover a single `scope', avoid grouping unrelated changes in a single commit;
+    \item While not always possible, commits should also be `atomic'. Instead of a partial change, the project is left in a coherent stage.
```
I'd rather say "coherent state" rather than "coherent stage". Previous slide already introduced the concept of "staging changes", which doesn't need to be confused with what is meant here.

henriquesimoes commented on git-tutorial.tex:
```diff
@@ -93,7 +93,7 @@ \subsection{Commiting}
   \end{itemize}
   \begin{block}{}
-    \texttt{\$ git commit}
+    \texttt{\$ EDITOR=nano git commit}
```
As we discussed in person, `GIT_EDITOR` has precedence over the configuration file, which in turn has precedence over both `VISUAL` and `EDITOR` (see `git-commit(1)`). This means that recommending the usage of `EDITOR` is not sufficient to ensure that anyone attending the workshop will indeed use `nano`, as there might be the case where `~/.gitconfig` (or the local repo config) contains `core.editor` key set. On the other hand, using `GIT_EDITOR` will cause any other editor configuration to be ignored. This means anyone that has already chosen their favorite editor beforehand will have `nano` launched.

In any case, the editor choice is inconsistent with the example (and recommendation?) provided in the very first slide. IMHO, it should one or the other across the entire document.

Besides all of that, in my point of view, we should actually drop this commit. I think the intention in this command block is to show the everyday command that is used to commit a change. Indeed, that's pretty much everything I type when I want to commit something. I don't think anyone should jot down that they must export `EDITOR` inline every time they want to git-commit.

henriquesimoes commented on git-tutorial.tex:
```diff
@@ -100,6 +100,22 @@ \subsection{Commiting}
+\subsection{Commit messages}
+\begin{frame}{Commit messages}
+  \begin{itemize}
+    \item What's an \textbf{in-depth explanation}? The items below aren't an exhaustive list, but should be considered when writing;
```
IMO, this should out of the `itemize` block.

I haven't checked if this a pattern used elsewhere, but it could be inside a `block` with a header. Something like this:

```latex
\begin{block}{What's an \textbf{in-depth explanation?}}
  \begin{itemize}
    ...
  \end{itemize}
\end{block}
```

henriquesimoes commented on git-tutorial.tex:
```diff
@@ -100,6 +100,22 @@ \subsection{Commiting}
+    \item Explain \textbf{how} a change works: remember that what's obvious to you now might not be obvious 3 months from then;
+    \item Explain \textbf{why} this strategy was chosen: were other options evaluated? Which ones? What were the advantages of this one?
+    \item Mention relevant results: benchmarks, LoC reduction, etc;
+    \item Difficulties encountered, bugs in initial versions, attention points can also be mentioned.
```
Links to external references?

ericonr committed:
tutorial: recommed git restore --staged.

The official docs recommend git-restore(1) for everything it can do [1],
including "git restore --staged" to manipulate the index.

[1] https://github.com/git/git/commit/80f537f79c16efeb7b92b3409ede434a230b5679

ericonr committed:
tutorial: suggest using nano as editor.

It's a friendlier editor, if desired by users. Since vim will apply
basic word wrapping automatically for commits, we do the same with nano,
using the suggestions from [1].

[1] https://stackoverflow.com/a/42202178

henriquesimoes reviewed (approved):
That looks fine by me. ;)

augustofg commented on git-tutorial.tex:
```diff
@@ -99,6 +104,23 @@ \subsection{Commiting}
+    \item Difficulties encountered, bugs in initial versions, attention points can also be mentioned.
+  \end{itemize}
+  \begin{block}{}
+    Commit messages have formatting rules: the title should have 50 or fewer characters, and lines of the message body should be wrapped at 72 characters.
```
I would change this to `preferably, the title should have 50 or fewer characters. Lines of the message body should be wrapped at 72 characters.`

ericonr replied:
Done

augustofg reviewed (approved):
:+1:

gustavosr8 commented:
@ericonr WDYT about mention that changes should be described in imperative mood as if you are giving orders to the codebase to change its behavior ?

ericonr commented:
I think guides online calling it imperative is wrong. It's just using the verb without conjugation. In Portuguese, I wouldn't say "Conserte a falha de segmentação", I'd say "Consertar a falha de segmentação". The usual suggestion to check if your commit title makes sense is to say "If applied, this change will" and then the commit title, which works both for Portuguese and English if you use the infinitive; the fact that the imperative in English is the same as the infinitive is a coincidence.

It should also be applied only to the commit title, not to the rest of the message. And it should be followed along with the commit message style (e.g. "module: change x" or "feat(add feature)")

I will see if I can work that into the guide. The commit message slide is pretty long already, I might have to split it into two slides...
